CrossRefPlugin.php

```diff
@@ -162,6 +165,68 @@ public function addRegistrationAgencyName(string $hookName, array $args): bool {
         return HOOK::CONTINUE;
     }
 
+    /**
+     * Adds self to "allowed" list of pub object types that can be assigned DOIs for this registration agency.
+     *
+     * @param string $hookName DoiSetupSettingsForm::getObjectTypes
+     * @param array $args [
+     *      @option array &$objectTypeOptions
+     * ]
+     * @return bool
+     */
+    public function addAllowedObjectTypes(string $hookName, array $args): bool {
+        $objectTypeOptions = &$args[0];
+        $allowedTypes = $this->getAllowedDoiTypes();
+
+        $objectTypeOptions = array_map(function($option) use ($allowedTypes) {
+            if (in_array($option['value'], $allowedTypes)) {
+                $option['allowedBy'][] = $this->getName();
+            }
+            return $option;
+        }, $objectTypeOptions);
+
+        return Hook::CONTINUE;
+    }
+
+    /**
+     * Add validation rule to Context for restriction of allowed pubObject types for DOI registration.
+     *
+     * @param string $hookName
+     * @param array $args
+     * @return bool
+     * @throws \Exception
+     */
+    public function validateAllowedPubObjectTypes(string $hookName, array $args): bool
+    {
+        $errors = &$args[0];
+        $props = $args[2];
+
+        if (!isset($props['enabledDoiTypes'])) {
+            return Hook::CONTINUE;
+        }
+
+        $contextId = $props['id'];
+        if (empty($contextId)) {
+            throw new \Exception("A context ID must be present to edit context settings");
+        }
```

Comment:
That's a good question. I don't think it would be needed because this is always being called from an API context where that has already been checked. It doesn't look like any of the other `ContextService` validation checks check the context ID against the current context. If there's another case I'm not thinking of, we can always add it in.

Reply:
Yes, it seems 'id' is readonly i.e. it should not be able for the user to change the props['id']...